pydantic_graph/pydantic_graph/graph.py

```diff
@@ -0,0 +1,145 @@
+
+RunSignatureT = ParamSpec('RunSignatureT')
+RunEndT = TypeVar('RunEndT', default=None)
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
+
+
+@dataclass(init=False)
+class Graph(Generic[StateT, RunEndT]):
+    """Definition of a graph."""
+
+    name: str | None
+    nodes: tuple[type[BaseNode[StateT, RunEndT]], ...]
+    node_defs: dict[str, NodeDef[StateT, RunEndT]]
+
+    def __init__(
+        self,
+        *,
+        nodes: Sequence[type[BaseNode[StateT, RunEndT]]],
+        state_type: type[StateT] | None = None,
+        name: str | None = None,
+    ):
+        self.name = name
+
+        _nodes_by_id: dict[str, type[BaseNode[StateT, RunEndT]]] = {}
+        for node in nodes:
+            node_id = node.get_id()
+            if existing_node := _nodes_by_id.get(node_id):
+                raise GraphSetupError(f'Node ID `{node_id}` is not unique — found in {existing_node} and {node}')
+            else:
+                _nodes_by_id[node_id] = node
+        self.nodes = tuple(_nodes_by_id.values())
+
+        parent_namespace = get_parent_namespace(inspect.currentframe())
+        self.node_defs: dict[str, NodeDef[StateT, RunEndT]] = {
+            node.get_id(): node.get_node_def(parent_namespace) for node in self.nodes
+        }
+
+        self._validate_edges()
+
+    def _validate_edges(self):
+        known_node_ids = self.node_defs.keys()
+        bad_edges: dict[str, list[str]] = {}
+
+        for node_id, node_def in self.node_defs.items():
+            for edge in node_def.next_node_edges.keys():
+                if edge not in known_node_ids:
+                    bad_edges.setdefault(edge, []).append(f'`{node_id}`')
+
+        if bad_edges:
+            bad_edges_list = [f'`{k}` is referenced by {_utils.comma_and(v)}' for k, v in bad_edges.items()]
+            if len(bad_edges_list) == 1:
+                raise GraphSetupError(f'{bad_edges_list[0]} but not included in the graph.')
+            else:
+                b = '\n'.join(f' {be}' for be in bad_edges_list)
+                raise GraphSetupError(f'Nodes are referenced in the graph but not included in the graph:\n{b}')
+
+    async def next(
+        self, state: StateT, node: BaseNode[StateT, RunEndT], history: list[HistoryStep[StateT, RunEndT]]
+    ) -> BaseNode[StateT, Any] | End[RunEndT]:
+        node_id = node.get_id()
+        if node_id not in self.node_defs:
+            raise GraphRuntimeError(f'Node `{node}` is not in the graph.')
+
+        history_step: NodeEvent[StateT, RunEndT] | None = NodeEvent(state, node)
+        history.append(history_step)
+
+        ctx = GraphContext(state)
+        with _logfire.span('run node {node_id}', node_id=node_id, node=node):
+            start = perf_counter()
+            next_node = await node.run(ctx)
+            history_step.duration = perf_counter() - start
+        return next_node
+
+    async def run(
+        self,
+        state: StateT,
+        node: BaseNode[StateT, RunEndT],
```

Comment:
We can rename node to start_node in the future to indicate that this is really the beginning of the graph.

Reply:
Just a suggestion

Reply:
So I think if the parameter is named as start_node it will be clearer (more obvious) to the caller of the run() method that is the start of the graph

Reply:
thanks, fixed.